File: bnb/simplex.py
'''Pure Python implementation of simplex algorithm.'''

from tabulate import tabulate

class Tableau:
    '''Simplex tableau.

    Explicitly track simplex tableau.
    '''

    def __init__(
            self, c,
            A_ub=[], b_ub=[],
            A_eq=[], b_eq=[],
            A_lb=[], b_lb=[],
            disp=False):
        '''Solve the linear program.'''

        self.c = c
        self.A_ub = A_ub
        self.b_ub = b_ub
        self.A_lb = A_lb
        self.b_lb = b_lb
        self.A_eq = A_eq
        self.b_eq = b_eq
        self.disp = disp

        # Find any negative entries in b and make positive
        remove_idx = set()
        for ii, val in enumerate(self.b_ub):
            if val < 0:
                # Multiply the leq inequality by -1 turning it into geq
                self.A_lb.append([-1*v0 for v0 in self.A_ub[ii]])
                self.b_lb.append(-1*val)
                remove_idx.add(ii)
        self.A_ub = [row for ii, row in enumerate(self.A_ub) if ii not in remove_idx]
        self.b_ub = [val for ii, val in enumerate(self.b_ub) if ii not in remove_idx]

        remove_idx = set()
        for ii,  val in enumerate(self.b_lb):
            if val < 0:
                # turns into leq inequality
                self.A_ub.append([-1*v0 for v0 in self.A_lb[ii]])
                self.b_ub.append([-1*val])
                remove_idx.add(ii)
        self.A_lb = [row for ii, row in enumerate(self.A_lb) if ii not in remove_idx]
        self.b_lb = [val for ii, val in enumerate(self.b_lb) if ii not in remove_idx]

        for ii, val in enumerate(self.b_eq):
            if val < 0:
                # Remains an equality
                self.A_eq[ii] = [-1*v0 for v0 in self.A_eq[ii]]
                self.b_eq[ii] *= -1
        
        # Allocate the table
        self.num_vars = len(c)
        self.num_slack = len(b_ub)
        self.num_surplus = len(b_lb)
        self.num_artificial = len(b_lb) + len(b_eq)
        self.nbasic_var_labels = 1
        self.nrhs_col = 1
        self.nobjective_row = 1
        self.nhdr_row = 1
        self.nA_ub = len(b_ub)
        self.nA_lb = len(b_lb)
        self.nA_eq = len(b_eq)

        # Determine if we need one or two phase solution
        self.two_phase = self.num_artificial > 0
        
        self.num_cols = (
            self.num_vars +
            self.num_slack +
            self.num_surplus +
            self.num_artificial +
            self.nbasic_var_labels +
            self.nrhs_col
        )
        self.num_rows = (
            self.nhdr_row +
            self.nobjective_row +
            self.nA_ub +
            self.nA_lb +
            self.nA_eq
        )
        self.tableau = []
        for ii in range(self.num_rows):
            self.tableau.append([0]*self.num_cols)

        # A_ub gets a slack variable for each row
        # A_eq gets an artificial variable for each row
        # A_lb gets a surplus variable (negative slack variable) and artificial variable for each row

        # Fill in headers of the tableau
        self.init_hdr()

        # Fill in values
        self.init_tableau_values()

        # Solve phase 1 if we need to
        if self.two_phase:

            # Change the problem to minimize sum of artificial variables
            phase1_c = [''] + [0]*(self.num_vars+self.num_slack+self.num_surplus) + [1]*(self.num_artificial)

            # Add the new objective row to the top -- we now have two objective functions we're tracking!
            self.tableau.insert(self.nhdr_row, phase1_c + [0])
            self.nobjective_row += 1
            self.num_rows += 1

            # Since we have no negatives in the objective row, use artificial variable rows to
            # eliminate entries in the objective column
            offset = self.nbasic_var_labels + self.num_vars + self.num_slack + self.num_surplus
            for ii in range(self.num_artificial):

                # Find current artificial variable row
                for jj in range(self.nhdr_row+self.nobjective_row, self.num_rows):
                    if self.tableau[jj][offset+ii] == 1: # should be 1
                        row_idx = jj
                        break

                # Update basic variable labels
                self.tableau[row_idx][0] = self.tableau[0][offset+ii]

                # Remove entry from objective
                row_vals = self.tableau[row_idx]
                for jj in range(self.nbasic_var_labels, self.num_cols):
                    self.tableau[self.nhdr_row][jj] -= row_vals[jj]

            # Solve the phase 1 problem
            self.solve()

            # If we get an objective value other than 0, problem is infeasible
            if self.get_obj() != 0:
                raise ValueError('There is no solution!')

            # Now solve phase 2 by removing artificial variable columns and phase 1 objective row
            del self.tableau[self.nhdr_row]
            self.nobjective_row -= 1
            self.num_rows -= 1

            offset = self.nbasic_var_labels+self.num_vars+self.num_slack+self.num_surplus
            for ii in range(self.num_rows):
                del self.tableau[ii][offset:offset+self.num_artificial]
            self.num_cols -= self.num_artificial
            self.num_artificial = 0

            # Can solve the phase 2 problem as normal

        self.solve()
        
    def solve(self):
        '''Pivot till you can't pivot no more.'''
        # Start pivoting
        done = False
        while not done:
            done = self.pivot()

    def get_obj(self):
        '''Find the current objective function value.'''
        return self.tableau[self.nhdr_row][-1]
            
    def pivot(self, tol=1e-8):
        '''Perform a pivot.'''

        # Check out starting state
        if self.disp:
            print(self)
        
        # Check exit conditions
        obj_row = self.tableau[self.nhdr_row][self.nbasic_var_labels:-self.nrhs_col]
        # use tol to make sure value is not just slightly negative, i.e., basically 0
        is_neg = [val < 0 and abs(val) > tol for val in obj_row]
        if not any(is_neg):
            return True # we're done!

        # Choose incoming variable from strictly negative objective cols
        min_index, _min_value = min(enumerate(zip(obj_row, is_neg)), key=lambda x0: x0[1][0]*x0[1][1])
        pivot_col = min_index + self.nbasic_var_labels

        # Choose the outgoing variable using ratio test
        ratios = [
            (self.tableau[ii][-1]/self.tableau[ii][pivot_col], ii) for ii in range(
                self.nhdr_row+self.nobjective_row, self.num_rows) if self.tableau[ii][pivot_col] > 0]
        if len(ratios) == 0:
            print(self)
            raise ValueError('Unbounded solution!')
        _min_ratio, pivot_row = min(ratios, key=lambda x0: x0[0])

        # Update the basic variable labels with the incoming variable
        self.tableau[pivot_row][0] = self.tableau[0][pivot_col]
        
        # Pivot! Pivot row divided by pivot value
        pivot_val = self.tableau[pivot_row][pivot_col]
        for jj in range(self.nbasic_var_labels, self.num_cols):
            self.tableau[pivot_row][jj] /= pivot_val
        pivot_row_vals = self.tableau[pivot_row]

        # Now use pivot row to remove remaining entries in the pivot col
        pivot_col_vals = [(ii, self.tableau[ii][pivot_col]) for ii in range(self.nhdr_row, self.num_rows) if ii != pivot_row and abs(self.tableau[ii][pivot_col]) > 0]
        for ii, val in pivot_col_vals:
            for jj in range(self.nbasic_var_labels, len(self.tableau[ii])):
                self.tableau[ii][jj] -= pivot_row_vals[jj]*val

        # Tell caller we didn't detect exiting condition and pivoted
        return False
    
    def init_tableau_values(self):
        '''Fill in initial values for constraints.'''

        # header and objective rows come first along rows
        # basic variable headers come first along columns
        row_offset = self.nhdr_row + self.nobjective_row
        col_offset = self.nbasic_var_labels
        
        slack_idx = self.num_vars + col_offset
        surplus_idx = self.num_vars + self.num_slack + col_offset
        artificial_idx = self.num_vars + self.num_slack + self.num_surplus + col_offset

        # Do objective row
        for ii, val in enumerate(self.c):
            self.tableau[self.nhdr_row][ii+col_offset] = -1*val
        
        # leq constraints
        for ii, row in enumerate(self.A_ub):
            # Include slack variables to be appended on this row
            for jj, val in enumerate(row):
                self.tableau[ii+row_offset][jj+col_offset] = val
            self.tableau[ii+row_offset][slack_idx] = 1
            slack_idx += 1

            # Also get RHS values
            self.tableau[ii+row_offset][-1] = self.b_ub[ii]

        # eq constraints
        for ii, row in enumerate(self.A_eq):
            # Include artificial variable on this row
            for jj, val in enumerate(row):
                self.tableau[self.nA_ub+ii+row_offset][jj+col_offset] = val
            self.tableau[self.nA_ub+ii+row_offset][artificial_idx] = 1
            #self.tableau[self.nhdr_row][artificial_idx] = -1*self.bigM # if bigM
            artificial_idx += 1

            # Also get RHS values
            self.tableau[self.nA_ub+ii+row_offset][-1] = self.b_eq[ii]

        # geq constraints
        for ii, row in enumerate(self.A_lb):
            # Include suprplus and artificial variables on this row
            for jj, val in enumerate(row):
                self.tableau[self.nA_ub+self.nA_eq+ii+row_offset][jj+col_offset] = val
            self.tableau[self.nA_ub+self.nA_eq+ii+row_offset][artificial_idx] = 1
            #self.tableau[self.nhdr_row][artificial_idx] = -1*self.bigM # if bigM
            self.tableau[self.nA_ub+self.nA_eq+ii+row_offset][surplus_idx] = -1
            artificial_idx += 1
            surplus_idx += 1

            # Also get RHS values
            self.tableau[self.nA_ub+self.nA_eq+ii+row_offset][-1] = self.b_lb[ii]
                
    def init_hdr(self):
        '''Fill in initial values for tableau header.'''
        idx = 0

        # Column tracking basic variables is first
        self.tableau[0][idx] = 'basic'
        idx += 1
        
        # Start with variables that appear in objective function,
        # then do slack, surplus, and artificial
        for ii in range(self.num_vars):
            self.tableau[0][idx] = 'x' + str(ii)
            idx += 1
        for ii in range(self.num_slack):
            self.tableau[0][idx] = 'sl' + str(ii)
            idx += 1
        for ii in range(self.num_surplus):
            self.tableau[0][idx] = 'su' + str(ii)
            idx += 1
        for ii in range(self.num_artificial):
            self.tableau[0][idx] = 'a' + str(ii)
            idx += 1
        # End with RHS
        self.tableau[0][idx] = 'RHS'
        idx += 1

        # Did we get all of them?
        assert idx == self.num_cols

        # Now do row headers
        self.tableau[1][0] = '' # objective row, no basic var here
        
        # Get a feasible solution by setting slack to RHS for A_ub
        col_offset = self.nbasic_var_labels
        row_offset = self.nhdr_row + self.nobjective_row
        for jj in range(self.nA_ub):
            self.tableau[row_offset + jj][0] = self.tableau[0][self.num_vars + col_offset + jj]

        # Set artificial to RHS for A_eq (because surplus is negative)
        offset = self.num_vars + self.num_slack + self.num_slack + col_offset
        for jj in range(self.nA_eq):
            self.tableau[row_offset + self.nA_ub + jj][0] = self.tableau[0][offset + jj]

        # Set artificial to RHS for A_lb
        offset += self.nA_eq
        for jj in range(self.nA_lb):
            self.tableau[row_offset + self.nA_ub + self.nA_eq + jj][0] = self.tableau[0][offset + jj]

    def solution(self):
        '''Return all useful info about solution.'''

        fopt = self.get_obj()

        # Extract basic variables from tableau
        basic_vars = [r[0] for r in self.tableau[self.nhdr_row+self.nobjective_row:]]
        x = [0]*self.num_vars
        s = [0]*(self.num_slack+self.num_surplus) # consider surplus variables as slack
        for ii, v in enumerate(basic_vars):
            if 'x' in v:
                idx = int(v[1:])
                x[idx] = self.tableau[self.nhdr_row+self.nobjective_row+ii][-1]
            elif 'sl' in v:
                idx = int(v[2:])
                s[idx] = self.tableau[self.nhdr_row+self.nobjective_row+ii][-1]
            elif 'su' in v:
                idx = int(v[2:])
                s[idx+self.num_slack] = self.tableau[self.nhdr_row+self.nobjective_row+ii][-1]
            else:
                raise ValueError('Got unexpected basic variable label: %s' % v)

        # Extract the dual solution from the tableau
        # TODO: I only know how to do this for symmetric dual problems (only A_ub constraints)
        # TODO: Seems like it works for A_lb constraints as well?
        dual = None
        if self.num_artificial == 0:# and self.num_surplus == 0:
            # Dual values are in the objective col for slack vars
            dual = self.tableau[self.nhdr_row][self.nbasic_var_labels+self.num_vars:-1]

        return {
            'fopt': fopt,
            'x': x,
            'slack': s,
            'dual': dual,
        }
            
    def __repr__(self):        
        return '\n' + tabulate(
            self.tableau,
            headers="firstrow",
            tablefmt='orgtbl',
            floatfmt=".1f") + '\n'

def simplex(
        c,
        A_ub=[], b_ub=[],
        A_eq=[], b_eq=[],
        A_lb=[], b_lb=[],
        disp=False):
    '''Solve a linear program using the simplex method.

    Parameters
    ----------
    c : list
        Objective coefficients.
    A_ub, b_ub : list of lists, list
        Less than or equal constraints.
    A_eq, b_eq : list of lists, list
        Equality constraints.
    A_lb, b_lb : list of lists, list
        Greater than or equal constraints.
    disp : bool, optional
        Show the tableau at each pivot.

    Returns
    -------
    res : dict
    
        - x : list
            Solution for the variables that maximize c @ x.
        - slack : list
            Slack variable values at the optimum.
        - fopt : float
            Objective value at the optimum.
        - dual : list or None
            Dual variable values at the optimum. These are provided
            only when there are no equality constraints, i.e.,
            A_eq=[] and b_eq=[].

    Notes
    -----
    Explicitly enumerates tableau.

    Will transform all inequality and equality constraints into
    standard form and solve using the two phase simplex algorithm.

    Assumes all variables are non-negative.
    '''

    t = Tableau(c, A_ub, b_ub, A_eq, b_eq, A_lb, b_lb, disp=disp)
    return t.solution()

if __name__ == '__main__':
    pass

File: src/simplex.cpp
#include <vector>
#include <stack>
#include <string>
#include <algorithm>
#include <functional>
#include <iterator>
#include <assert.h> // assert
#include <iostream>

#include "simplex.hpp"

namespace simplex {

  template<class T>
  Simplex<T>::Simplex(const std::vector<T> c_,
		      const std::vector<std::vector<T> > A_ub_, const std::vector<T> b_ub_,
		      const std::vector<std::vector<T> > A_eq_, const std::vector<T> b_eq_,
		      const std::vector<std::vector<T> > A_lb_, const std::vector<T> b_lb_,
		      const bool solve_dual_)
  {

    // May be computationally advantageous to solve dual instead of primal
    this->solve_dual = solve_dual_;
    assert(!this->solve_dual); // we don't do this yet
    
    // Store copy of coefficients and constraints
    this->c = std::vector<T>(c_);
    this->A_ub = std::vector<std::vector<T> >(A_ub_);
    this->A_eq = std::vector<std::vector<T> >(A_eq_);
    this->A_lb = std::vector<std::vector<T> >(A_lb_);
    this->b_ub = std::vector<T>(b_ub_);
    this->b_eq = std::vector<T>(b_eq_);
    this->b_lb = std::vector<T>(b_lb_);

    // Go ahead and make RHS all non-negative. No longer trust function
    // arguments to have correct constraints after this! Use Simplex::A_ub, etc.
    this->make_rhs_nonnegative();
    
    this->num_vars = this->c.size();
    this->num_slack = this->b_ub.size();
    this->num_surplus = this->b_lb.size();
    this->num_artificial = this->num_surplus + this->b_eq.size();

    // We have a two phase problem if we got artificial variables
    this->two_phase = this->num_artificial > 0;
    
    // Table specific stuff
    this->num_obj_rows = 1;
    
    // See if we need to run two phases
    this->two_phase = this->num_artificial > 0;
    
    // Initialize the tableau to all 0s
    this->allocate_tableau();

    // Make headers for the tableau
    this->make_hdrs();

    // Maybe fill in tableau before FBS?
    this->fill_initial_tableau();
    
    // Get an intitial feasible solution
    this->label_initial_fbs();
  }

  template<class T>
  void Simplex<T>::solve() {

    // Solve the phase 1 problem if we need to
    if (this->two_phase) {
      
      // Change the problem to minimize sum of artificial variables
      auto phase1_c = std::vector<T>(this->num_cols());
      std::fill_n(std::next(phase1_c.begin(), this->artificial_col_start_idx()), this->num_artificial, 1);

      // Add the new objective row to the top -- we now have two objective functions we're tracking!
      this->tableau.insert(this->tableau.cbegin(), phase1_c);
      this->num_obj_rows++;
      this->basis.insert(this->basis.cbegin(), "phase1");

      // Since we have no negatives in the objective row, use artificial variable rows to
      // eliminate entries in the objective column
      auto pivot_row = this->num_rows() - this->num_artificial;
      for (auto ii = this->artificial_col_start_idx(); ii < this->rhs_col_idx(); ++ii) {
	this->pivot(pivot_row, ii);
	pivot_row++;
      }

      // Solve the phase 1 problem
      this->two_phase = false;
      this->solve();

      // If we get an objective value other than 0, problem is infeasible
      if (this->get_obj_val() != 0) {
	std::cout << "There is no solution!" << std::endl;
	return;
      }

      // Now solve phase 2 by removing artificial variable columns and phase 1 objective row
      this->tableau.erase(this->tableau.cbegin());
      this->basis.erase(this->basis.cbegin());
      this->num_obj_rows--;
      for (auto & row : this->tableau) {
	row.erase(std::next(row.cbegin(), this->rhs_col_idx() - this->num_artificial),
		  std::prev(row.cend(), 1));
      }
      this->hdrs.erase(std::next(this->hdrs.cbegin(), this->rhs_col_idx() - this->num_artificial + this->hdr_var_start_idx()),
		       std::prev(this->hdrs.cend(), 1));
      this->num_artificial = 0;

      // Can solve the phase 2 problem as normal
    }
    
    // Keep pivoting till we can't pivot no more!
    // break to get out of loop
    while (true) {

      // Take a look
      this->show();

      // Check exit conditions
      auto obj_row_idx = this->obj_row_start_idx();
      auto & tol = this->tolerance;
      if (std::all_of(this->tableau[obj_row_idx].cbegin(),
		      std::prev(this->tableau[obj_row_idx].cend(), 1), // don't include RHS
		      [tol](const auto & el) {
			return el >= 0 || -1*el < tol;
		      })) {
	// All objective row elements are positive, we're done!
	break;
      }
		      
      // Choose incoming variable from strictly negative objective cols
      auto min_it = std::min_element(this->tableau[obj_row_idx].cbegin(),
				     std::prev(this->tableau[obj_row_idx].cend(), 1)); // don't include RHS
      assert(*min_it < 0);
      auto pivot_col = std::distance(this->tableau[obj_row_idx].cbegin(), min_it);
      
      // Choose the outgoing variable using ratio test.
      // If we store references to column major tableau we could get these easier
      auto ratios = std::vector<std::pair<std::size_t, T> >();
      auto rhs_col_idx = this->rhs_col_idx();
      std::size_t idx = 0;
      for (const auto & row : this->tableau) {
	if (row[pivot_col] > 0) {
	  ratios.push_back(std::make_pair(idx, row[rhs_col_idx]/row[pivot_col]));
	}
	idx++;
      }
      if (ratios.size() == 0) {
	std::cout << "Unbounded solution!" << std::endl;
	break;
      }
      auto ratio_min_it = std::min_element(ratios.cbegin(),
					   ratios.cend(),
					   [](const auto & a, const auto & b) -> bool {
					     return std::get<1>(a) < std::get<1>(b);
					   });
      auto pivot_row = std::get<0>(*ratio_min_it);
      
      // Do the pivot
      this->pivot(pivot_row, pivot_col);
    }
  }
  
  template<class T>
  void Simplex<T>::pivot(const std::size_t pivot_row, const std::size_t pivot_col) {

    // Update the basic variable labels with the incoming variable
    this->basis[pivot_row] = this->hdrs[this->hdr_var_start_idx() + pivot_col];
   
    // Pivot! Pivot row divided by pivot value
    auto pivot_val = this->tableau[pivot_row][pivot_col];
    std::transform(this->tableau[pivot_row].cbegin(),
		   this->tableau[pivot_row].cend(),
		   this->tableau[pivot_row].begin(),
		   [pivot_val](const auto & el) {
		     return el/pivot_val;
		   });

    // Now use pivot row to remove remaining entries in the pivot col
    for (std::size_t ii = 0; ii < this->num_rows(); ++ii) {
      if (ii == pivot_row) {
	continue;
      }
      auto val = this->tableau[ii][pivot_col];
      zip2(this->tableau[pivot_row].cbegin(),
	   this->tableau[pivot_row].cend(),
	   this->tableau[ii].begin(),
	   [val](const auto & el1, auto & el2) {
	     el2 -= val*el1;
	   });
    }    
  }

  template<class T>
  inline std::size_t Simplex<T>::hdr_var_start_idx() {
    return 1; // 'basic' is first entry
  }

  template<class T>
  inline std::size_t Simplex<T>::hdr_slack_start_idx() {
    return this->hdr_var_start_idx() + this->num_vars;
  }

  template<class T>
  inline std::size_t Simplex<T>::hdr_surplus_start_idx() {
    return this->hdr_slack_start_idx() + this->num_slack;
  }
  
  template<class T>
  inline std::size_t Simplex<T>::hdr_artificial_start_idx() {
    return this->hdr_surplus_start_idx() + this->num_surplus;
  }
  
  template<class T>
  void Simplex<T>::label_initial_fbs() {

    // First entry of the basis labels is reserved for objective row
    this->basis.push_back("obj");
    
    // Get a feasible solution by setting slack to RHS for A_ub
    std::copy_n(std::next(this->hdrs.cbegin(), this->hdr_slack_start_idx()),
		this->num_slack,
		std::back_inserter(this->basis));

    // Set artificial to RHS for A_eq (because surplus is negative)
    std::copy_n(std::next(this->hdrs.cbegin(), this->hdr_artificial_start_idx()),
		this->num_surplus,
		std::back_inserter(this->basis));

    // Set artificial to RHS for A_lb
    std::copy_n(std::next(this->hdrs.cbegin(), this->hdr_artificial_start_idx() + this->num_surplus),
		this->num_artificial - this->num_surplus,
		std::back_inserter(this->basis));

  }
  
  template<class T>
  void Simplex<T>::fill_initial_tableau() {

    // Do objective row
    auto obj_row_idx = this->obj_row_start_idx();
    std::copy(this->c.cbegin(), this->c.cend(), this->tableau[obj_row_idx].begin());
    transform_n(this->tableau[obj_row_idx].cbegin(),
		this->num_vars,
		this->tableau[obj_row_idx].begin(),
		std::negate<T>());

    // leq constraints
    auto slack_col = this->slack_col_start_idx();
    auto rhs_col = this->rhs_col_idx();
    zip3(this->A_ub.cbegin(),
	 this->A_ub.cend(),
	 this->b_ub.cbegin(),
	 std::next(this->tableau.begin(), this->A_ub_row_start_idx()),
	 [&slack_col, rhs_col](const auto & Ael, const auto & bel, auto & tableau_row) {

	   // Copy coefficients from A_ub to tableau
	   std::copy(Ael.cbegin(), Ael.cend(), tableau_row.begin());

	   // Include slack variable for each constraint
	   tableau_row[slack_col] = 1;

	   // Copy RHS value for this constraint
	   tableau_row[rhs_col] = bel;

	   // Move to next slack col
	   slack_col++;
	 });

    // eq constraints
    auto artificial_col = this->artificial_col_start_idx();
    zip3(this->A_eq.cbegin(),
	 this->A_eq.cend(),
	 this->b_eq.cbegin(),
	 this->tableau.begin() + this->A_eq_row_start_idx(),
	 [&artificial_col, rhs_col](const auto & Ael, const auto & bel, auto & tableau_row) {

	   // Copy coefficients from A_eq to tableau
	   std::copy(Ael.cbegin(), Ael.cend(), tableau_row.begin());

	   // Include artificial variable for each constraint
	   tableau_row[artificial_col] = 1;

	   // Copy RHS value for this constraint
	   tableau_row[rhs_col] = bel;

	   // Increment artificial col
	   artificial_col++;
	 });

    // geq constraints
    auto surplus_col = this->surplus_col_start_idx();
    zip3(this->A_lb.cbegin(),
	 this->A_lb.cend(),
	 this->b_lb.cbegin(),
	 this->tableau.begin() + this->A_lb_row_start_idx(),
	 [&surplus_col, &artificial_col, rhs_col](const auto & Ael, const auto & bel, auto & tableau_row) {

	   // Copy coefficients from A_eq to tableau
	   std::copy(Ael.cbegin(), Ael.cend(), tableau_row.begin());

	   // Include surplus and artificial variable
	   tableau_row[surplus_col] = -1;
	   tableau_row[artificial_col] = 1;

	   // Copy RHS value for this constraint
	   tableau_row[rhs_col] = bel;

	   // Increment cols
	   surplus_col++;
	   artificial_col++;
	 });

  }

  template<class T>
  void Simplex<T>::make_hdrs() {

    // Column tracking basic variables is first
    this->hdrs.push_back("basic");
        
    // Start with variables that appear in objective function,
    // then do slack, surplus, and artificial
    for (std::size_t ii = 0; ii < this->num_vars; ++ii) {
      this->hdrs.push_back("x" + std::to_string(ii));
    }
    for (std::size_t ii = 0; ii < this->num_slack; ++ii) {
      this->hdrs.push_back("sl" + std::to_string(ii));
    }
    for (std::size_t ii = 0; ii < this->num_surplus; ++ii) {
      this->hdrs.push_back("su" + std::to_string(ii));
    }
    for (std::size_t ii = 0; ii < this->num_artificial; ++ii) {
      this->hdrs.push_back("a" + std::to_string(ii));
    }
    // End with RHS
    this->hdrs.push_back("RHS");

    // Did we get all of them?
    assert(this->hdrs.size() - this->num_basic_label_cols == this->num_cols());
  }
  
  template<class T>
  void Simplex<T>::make_rhs_nonnegative() {
    // Find any negative entries in b and make positive
    auto remove_idx = std::stack<std::size_t>();
    std::size_t ii = 0;
    for (auto val : this->b_ub) {
      if (val < 0) {
	// Multiply the leq inequality by -1 turning it into geq
	auto r = std::vector<T>(this->A_ub[ii]);
	std::transform(r.cbegin(),
		       r.cend(),
		       r.begin(),
		       std::negate<T>());
	this->A_lb.push_back(r);
	this->b_lb.push_back(-1*val);
	remove_idx.push(ii);
      }
      ii += 1;
    }
    // Remove entries marked for deletion;
    // use stack so indices aren't messed up!
    while (!remove_idx.empty()) {
      auto row_idx = remove_idx.top();
      this->A_ub[row_idx] = this->A_ub.back();
      this->A_ub.pop_back();
      this->b_ub[row_idx] = this->b_ub.back();
      this->b_ub.pop_back();
      remove_idx.pop();
    }

    // Reset index counter and do the same for geq constraints
    ii = 0;
    for (auto val : this->b_lb) {
      if (val < 0) {
	// turns into leq inequality
	auto r = std::vector<T>(this->A_lb[ii]);
	std::transform(r.cbegin(),
		       r.cend(),
		       r.begin(),
		       std::negate<T>());
	this->A_ub.push_back(r);
	this->b_ub.push_back(-1*val);
	remove_idx.push(ii);
      }
    }
    while (!remove_idx.empty()) {
      auto row_idx = remove_idx.top();
      this->A_lb[row_idx] = this->A_lb.back();
      this->A_lb.pop_back();
      this->b_lb[row_idx] = this->b_lb.back();
      this->b_lb.pop_back();
      remove_idx.pop();
    }

    // Equality constraints just need to be negated
    ii = 0;
    for (auto val : this->b_eq) {
      if (val < 0) {
	// Remains an equality
	std::transform(this->A_eq[ii].cbegin(),
		       this->A_eq[ii].cend(),
		       this->A_eq[ii].begin(),
		       std::negate<T>());
	this->b_eq[ii] *= -1;
      }
    }
  }
  
  template<class T>
  void Simplex<T>::allocate_tableau() {
    this->tableau.resize(this->num_rows(), std::vector<T>(this->num_cols()));
  }

  template<class T>
  inline std::size_t Simplex<T>::num_rows() {
    return this->num_obj_rows +
      this->b_ub.size() +
      this->b_eq.size() +
      this->b_lb.size();
  }

  template<class T>
  inline std::size_t Simplex<T>::num_cols() {
    return this->num_vars +
      this->num_slack +
      this->num_surplus +
      this->num_artificial +
      this->num_rhs_cols;
  }

  template<class T>
  inline std::size_t Simplex<T>::obj_row_start_idx() {
    return 0; // Always start at beginning
  }

  template<class T>
  inline std::size_t Simplex<T>::phase1_obj_row_idx() {
    return this->obj_row_start_idx() + 1; // always the second obj
  }

  template<class T>
  inline std::size_t Simplex<T>::rhs_col_idx() {
    return this->num_cols() - 1;
  }

  template<class T>
  inline T Simplex<T>::get_obj_val() {
    return this->tableau[this->obj_row_start_idx()][this->rhs_col_idx()];
  }

  template<class T>
  inline T Simplex<T>::get_phase1_obj_val() {
    return this->tableau[this->phase1_obj_row_idx()][this->rhs_col_idx()];
  }
  
  template<class T>
  inline std::size_t Simplex<T>::A_ub_row_start_idx() {
    return this->obj_row_start_idx() + this->num_obj_rows;
  }

  template<class T>
  inline std::size_t Simplex<T>::A_eq_row_start_idx() {
    return this->A_ub_row_start_idx() + this->b_ub.size();
  }

  template<class T>
  inline std::size_t Simplex<T>::A_lb_row_start_idx() {
    return this->A_eq_row_start_idx() + this->b_eq.size();
  }

  template<class T>
  inline std::size_t Simplex<T>::slack_col_start_idx() {
    return this->num_vars;
  }

  template<class T>
  inline std::size_t Simplex<T>::surplus_col_start_idx() {
    return this->slack_col_start_idx() + this->num_slack;
  }
  
  template<class T>
  inline std::size_t Simplex<T>::artificial_col_start_idx() {
    return this->surplus_col_start_idx() + this->num_surplus;
  }

  template<class T>
  inline std::size_t Simplex<T>::slack_col_by_A_ub_row(std::size_t row_idx) {
    return this->slack_col_start_idx() + row_idx;
  }

  template<class T>
  void Simplex<T>::show() {

    // Spit out header
    for (const auto & el : this->hdrs) {
      std::cout << el << ' ';
    }
    std::cout << '\n';

    // Spit out table underneath
    zip2(this->tableau.cbegin(),
	 this->tableau.cend(),
	 this->basis.cbegin(),
	 [](const auto & row, const auto & row_hdr) {
	   std::cout << row_hdr << ' ';
	   for (const auto & col : row) {
	     std::cout << col << ' ';
	   }
	   std::cout << '\n';
	 });
    std::cout << std::endl; // flush
  }

  // Explicit template instantiations to make sure we only work with numeric types
  template class Simplex<float>;
  template class Simplex<double>;
  template class Simplex<long double>;

}

/*
int main() {

  auto c = std::vector<double>({3, 5});

  auto A_ub = std::vector<std::vector<double> >();
  A_ub.push_back(std::vector<double>({0, 1}));
  auto b_ub = std::vector<double>({6});

  auto A_eq = std::vector<std::vector<double> >();
  A_eq.push_back(std::vector<double>({3, 2}));
  auto b_eq = std::vector<double>({18});

  auto A_lb = std::vector<std::vector<double> >();
  A_lb.push_back(std::vector<double>({3, 5}));
  auto b_lb = std::vector<double>({2});

  auto s = simplex::Simplex<double>(c,
				    A_ub, b_ub,
				    A_eq, b_eq,
				    A_lb, b_lb,
				    false);

  s.show();

  s.solve();
  
  return 0;
}
*/
